python_modules/dagster/dagster/_daemon/sensor.py

```diff
@@ -957,6 +980,20 @@ def _handle_run_requests(
     submit_threadpool_executor: Optional[ThreadPoolExecutor],
     sensor_debug_crash_flags: Optional[SingleInstigatorDebugCrashFlags] = None,
 ):
+    # first, write out any evaluations without any run ids
```

Comment:
this is exactly like the asset daemon currently i assume, i remember the ordering there being tricky

Reply:
yep I mostly cargo-culted your work here, which I spent a good amount of time thinking about and feel comfortable with. the important invariants here are:

1. the new cursor is written to the db at the exact time as the run ids are reserved and the run requests are stored. this means that at any point where we generate a new cursor, we know the exact set of work we need to kick off in order for the cursor to be advanced in a valid way, meaning we can block the next evaluation until all of the required runs are submitted.
2. before the cursor et. al. are written to the db, we write the evaluations. This means that if we ever kick off work, we are guaranteed to have a record of why that work was kicked off.